cdc/owner/async_sink.go

```diff
@@ -0,0 +1,181 @@
+	cerror "github.com/pingcap/ticdc/pkg/errors"
+	"github.com/pingcap/ticdc/pkg/filter"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultErrChSize = 1024
+)
+
+// AsyncSink is a async sink design for owner
+type AsyncSink interface {
+	Initialize(ctx cdcContext.Context, tableInfo []*model.SimpleTableInfo) error
+	EmitCheckpointTs(ctx cdcContext.Context, ts uint64)
+	// EmitDDLEvent emits DDL event asynchronously and retuen true if the DDL is executed
+	EmitDDLEvent(ctx cdcContext.Context, ddl *model.DDLEvent) (bool, error)
+	SinkSyncpoint(ctx cdcContext.Context, checkpointTs uint64) error
+	Close() error
+}
+
+type asyncSinkImpl struct {
+	sink           sink.Sink
+	syncpointStore sink.SyncpointStore
+
+	checkpointTs model.Ts
+
+	ddlCh         chan *model.DDLEvent
+	ddlFinishedTs model.Ts
+	ddlSentTs     model.Ts
+
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+	errCh  chan error
+}
+
+func newAsyncSink(ctx cdcContext.Context) (AsyncSink, error) {
+	ctx, cancel := cdcContext.WithCancel(ctx)
+	changefeedID := ctx.ChangefeedVars().ID
+	changefeedInfo := ctx.ChangefeedVars().Info
+	filter, err := filter.NewFilter(changefeedInfo.Config)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	errCh := make(chan error, defaultErrChSize)
+	s, err := sink.NewSink(ctx, changefeedID, changefeedInfo.SinkURI, filter, changefeedInfo.Config, changefeedInfo.Opts, errCh)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	asyncSink := &asyncSinkImpl{
+		sink:   s,
+		ddlCh:  make(chan *model.DDLEvent, 1),
+		errCh:  errCh,
+		cancel: cancel,
+	}
+	if changefeedInfo.SyncPointEnabled {
+		asyncSink.syncpointStore, err = sink.NewSyncpointStore(ctx, changefeedID, changefeedInfo.SinkURI)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if err := asyncSink.syncpointStore.CreateSynctable(ctx); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	asyncSink.wg.Add(1)
+	go asyncSink.run(ctx)
+	return asyncSink, nil
+}
+
+func (s *asyncSinkImpl) Initialize(ctx cdcContext.Context, tableInfo []*model.SimpleTableInfo) error {
+	return s.sink.Initialize(ctx, tableInfo)
+}
+
+func (s *asyncSinkImpl) run(ctx cdcContext.Context) {
+	defer s.wg.Done()
+	// TODO make the tick duration configurable
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	var lastCheckpointTs model.Ts
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case err := <-s.errCh:
+			ctx.Throw(err)
+			return
+		case <-ticker.C:
+			checkpointTs := atomic.LoadUint64(&s.checkpointTs)
+			if checkpointTs == 0 || checkpointTs <= lastCheckpointTs {
+				continue
+			}
+			lastCheckpointTs = checkpointTs
+			if err := s.sink.EmitCheckpointTs(ctx, checkpointTs); err != nil {
+				ctx.Throw(errors.Trace(err))
+				return
+			}
+		case ddl := <-s.ddlCh:
+			err := s.sink.EmitDDLEvent(ctx, ddl)
+			failpoint.Inject("InjectChangefeedDDLError", func() {
+				err = cerror.ErrExecDDLFailed.GenWithStackByArgs()
+			})
+			if err == nil {
```

Comment:
```
if err == nil {
				log.Info("Execute DDL succeeded", zap.String("changefeed", ctx.ChangefeedVars().ID), zap.Reflect("ddl", ddl))
				atomic.StoreUint64(&s.ddlFinishedTs, ddl.CommitTs)
			} else if cerror.ErrDDLEventIgnored.Equal(errors.Cause(err)) {
				// If DDL executing failed, and the error can be ignored, mark the ddl is executed successfully and print a log
				log.Info("Execute DDL ignored", zap.String("changefeed", ctx.ChangefeedVars().ID), zap.Reflect("ddl", ddl))
				atomic.StoreUint64(&s.ddlFinishedTs, ddl.CommitTs)
			}
```
seems the statement looks exactly same besides the loginfo _**succeeded vs ignored**_